### PR TITLE
chore(suite): fiat rates to common

### LINKED ISSUE #8110: Refactor fiat rates fetching logic
Integrate changes done in https://github.com/trezor/trezor-suite/pull/7960 into @suite-common/fiat-services and make it again shared logic between desktop and native. 

Discuss with @Nodonisko what issues he has been facing when he initially tried to do it shared right away.

If it's possible to do it gradually, please split it in more issues/PRs.

Maybe it can fix https://github.com/trezor/trezor-suite/issues/5089
It should definitelly decrease and optimize the number of requests on fiat rates.

### PR DESCRIPTION
**Pull request by AdamSchinzel**
Moved `@suite-native/fiat-rates` to `@suite-comon/wallet-core` and fixed showing ERC20 tokens fiat value

## Description

- Desktop now uses the same fiat rates fetching logic as mobile which should be more effective
- Replaced all access to state `state.wallet.fiat.coins` with `selectCoinsLegacy` selector
- Refactor `@suite-comon/wallet-types`
- Refactor `@suite-comon/wallet-utils`
- Value of ERC20 tokens is now visible and if there is still no value I added `NoRatesTooltip`
- Storage migration

## Related Issue

Resolve #8110 #9163 #8109

## Screenshots:
Screenshot 2023-08-17 at 1 54 48 PM

### CODE REVIEW

File: packages/suite/src/actions/suite/storageActions.ts [hunk 1]
```diff
@@ -20,6 +20,8 @@ import type { PreloadStoreAction } from 'src/support/suite/preloadStore';
 import { GraphData } from 'src/types/wallet/graph';
 import { deviceGraphDataFilterFn } from 'src/utils/wallet/graph';
 import { selectCoinjoinAccountByKey } from 'src/reducers/wallet/coinjoinReducer';
+import { selectCoinsLegacy } from '@suite-common/wallet-core';
+import { useSelector } from 'src/hooks/suite';
```
Comment (matejkriz) [inline packages/suite/src/actions/suite/storageActions.ts]: I don't think we can use redux useSelector hook outside components in action creator, can we?

File: packages/suite/src/middlewares/wallet/storageMiddleware.ts [hunk 2]
```diff
@@ -64,19 +63,6 @@ const storageMiddleware = (api: MiddlewareAPI<Dispatch, AppState>) => {
...
-
-            if (fiatRatesActions.removeFiatRate.match(action)) {
-                api.dispatch(
-                    storageActions.removeFiatRate(
-                        action.payload.symbol,
-                        action.payload.tokenAddress,
-                    ),
-                );
-            }
-
```
Comment (matejkriz) [inline packages/suite/src/middlewares/wallet/storageMiddleware.ts]: Why is this removed?

Comment (matejkriz) [reply packages/suite/src/middlewares/wallet/storageMiddleware.ts]: also removeFiatRate in storageActions is no more used but not deleted

File: suite-common/wallet-core/src/fiat-rates/fiatRatesMiddleware.ts [hunk 3]
```diff
@@ -1,76 +1,39 @@
...
 } from './fiatRatesThunks';
 import { blockchainActions } from '../blockchain/blockchainActions';
-import { selectAccountTransactions } from '../transactions/transactionsReducer';
+import { accountsActions } from '../accounts/accountsActions';
+import { transactionsActions } from '../transactions/transactionsActions';
 
 export const prepareFiatRatesMiddleware = createMiddlewareWithExtraDeps(
     (action, { dispatch, extra, next, getState }) => {
         const {
+            actions: { setWalletSettingsLocalCurrency },
```
Comment (matejkriz) [inline suite-common/wallet-core/src/fiat-rates/fiatRatesMiddleware.ts]: if `changeWalletSettingsNetworks` is not used anymore, no need to keep it in extraDependencies

File: packages/suite/src/actions/suite/storageActions.ts (hunk 1 above)
Comment (matejkriz) [reply packages/suite/src/actions/suite/storageActions.ts]: nope image

File: suite-common/wallet-core/src/fiat-rates/fiatRatesReducer.ts [hunk 4]
```diff
@@ -1,91 +1,75 @@
...
+
+                const currentRate = state[rateType]?.[fiatRateKey];
+
+                // To prevent race condition someone will remove rate from state while fetching for example (during currency change etc.)
+                if (!currentRate) return;
+
+                state[rateType][fiatRateKey] = {
+                    ...currentRate,
+                    rate: action.payload,
+                    lastSuccessfulFetchTimestamp: Date.now() as Timestamp,
```
Comment (matejkriz) [inline suite-common/wallet-core/src/fiat-rates/fiatRatesReducer.ts]: reducer should be pure, so timestamp should go from action creator

Comment (matejkriz) [reply suite-common/wallet-core/src/fiat-rates/fiatRatesReducer.ts]: I see that this is copy-paste, but I believe it's wrong

File: packages/suite/src/middlewares/wallet/storageMiddleware.ts (hunk 2 above)
Comment (AdamSchinzel) [reply packages/suite/src/middlewares/wallet/storageMiddleware.ts]: We don't have `updateFiatRate` and `removeFiatRate` in `fiatRatesActions.ts` so we can't have it there anymore, was going to ask if this is something important or not?

Will remove these actions from these and the imports (should also solve the first thing in the review)

File: suite-common/wallet-core/src/fiat-rates/fiatRatesReducer.ts (hunk 4 above)
Comment (AdamSchinzel) [reply suite-common/wallet-core/src/fiat-rates/fiatRatesReducer.ts:52]: f0796072864f0ae73f74e1d235b43c4dc83a2dfb

File: suite-common/wallet-core/src/fiat-rates/fiatRatesMiddleware.ts (hunk 3 above)
Comment (AdamSchinzel) [reply suite-common/wallet-core/src/fiat-rates/fiatRatesMiddleware.ts]: f0796072864f0ae73f74e1d235b43c4dc83a2dfb

File: packages/suite/src/middlewares/wallet/storageMiddleware.ts (hunk 2 above)
Comment (AdamSchinzel) [reply packages/suite/src/middlewares/wallet/storageMiddleware.ts]: f0796072864f0ae73f74e1d235b43c4dc83a2dfb

File: packages/suite/src/actions/suite/storageActions.ts (hunk 1 above)
Comment (AdamSchinzel) [reply packages/suite/src/actions/suite/storageActions.ts]: f0796072864f0ae73f74e1d235b43c4dc83a2dfb

Comment (matejkriz) [reply packages/suite/src/actions/suite/storageActions.ts]: it could be a fixup if it fixes a specific commit...

File: packages/suite/src/middlewares/wallet/storageMiddleware.ts (hunk 2 above)
Comment (matejkriz) [reply packages/suite/src/middlewares/wallet/storageMiddleware.ts]: saveFiatRates still called from storageMiddleware...

Comment (matejkriz) [reply packages/suite/src/middlewares/wallet/storageMiddleware.ts]: I guess it was important, it stored downloaded fiat rates persistently in storage, so user don't have to download them all again on app reload/start. Not sure if some equivalent is needed or not. But even if not, this would need storage migration - no more needed data should be removed from the storage.

File: packages/suite/src/components/suite/FiatValue.tsx [hunk 5]
```diff
@@ -76,13 +76,13 @@ export const FiatValue = ({
 }: FiatValueProps) => {
     const { FiatAmountFormatter } = useFormatters();
     const localCurrency = useSelector(selectLocalCurrency);
-    const coins = useSelector(selectCoins);
+    const coins = useSelector(selectCoinsLegacy);
```
Comment (matejkriz) [inline packages/suite/src/components/suite/FiatValue.tsx]: Do we still need this legacy selector? Wouldn't be easier to use selectFiatRatesByFiatRateKey instead directly? Maybe not, I don't know.

File: packages/suite/src/middlewares/wallet/storageMiddleware.ts (hunk 2 above)
Comment (AdamSchinzel) [reply packages/suite/src/middlewares/wallet/storageMiddleware.ts]: As we discussed on Friday, migration will solve this (aa141d33ce5dcf686c307a780d6d3382bee8f1ec)

**Review by Nodonisko — COMMENTED** [review]
Briefly checked this PR and it seems good. Please test properly currency switching and all behavior around it because now rates are fetched only for currently active currency (except transactions).

File: suite-common/wallet-core/src/fiat-rates/fiatRatesReducer.ts (hunk 4 above)
Comment (AdamSchinzel) [reply suite-common/wallet-core/src/fiat-rates/fiatRatesReducer.ts]: 51f14248958e3f26c7955cac8548c8e16000a7a4

**Comment by AdamSchinzel** [comment]
Was testing it now for a quite some time and seems alright but can I ask @trezor/qa to try it even before this will be merged?

**Comment by AdamSchinzel** [comment]
Everything works for me when I tested it now but since there are quite a lot of changes can I ask @trezor/qa to test it even before merge, please? Thank you guys

**Comment by bosomt** [comment]
**QA OK**

image

image

image

image

**Info:**
 - Suite version: web 23.9.0 (f58d159ff47d8f6776ba63c16468b7a8078f7dea)
 - Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/116.0
 - OS: MacIntel
 - Screen: 1824x1140
 - Device: Trezor T2T1 2.6.1 regular (revision b151a3db6b04d2bee0bb68470e57b822443bf83d)
 - Transport: BridgeTransport 2.0.33

**Comment by bosomt** [comment]
And other fiat rates than ERC20 works as expected

image

**Comment by PeKne** [comment]
@bosomt This PR also influences the mobile version. Could you please briefly check that Trezor Suite mobile also works as expected? 🙏

**Comment by bosomt** [comment]
@PeKne as soon it gets merged to dev master i will retest it on  mobile app build 🙏🏿

File: suite-common/wallet-core/src/fiat-rates/fiatRatesReducer.ts (hunk 4 above)
Comment (matejkriz) [reply suite-common/wallet-core/src/fiat-rates/fiatRatesReducer.ts]: cool, thanks, just to be sure - why are you removing the default values for `rateType`? Is it because it was redundantly on all places? so `current` is still the default?

Comment (AdamSchinzel) [reply suite-common/wallet-core/src/fiat-rates/fiatRatesReducer.ts]: Yes, when I went though all places I found out that it was not necessary to have it there.

**Comment by szymonlesisz** [comment]
regarding last commit im afraid that you will need to rebase everyting, [there was an update](https://github.com/trezor/trezor-suite/pull/9198) of useSendForm and useRbfForm tests (using react toolkit)

**Comment by AdamSchinzel** [comment]
Upps didn't want to merge it...

**Comment by AdamSchinzel** [comment]
Upps mistake will fix it...
